OpenSim/Common/XsensDataReader.cpp

```diff
@@ -0,0 +1,224 @@
+#include "Simbody.h"
+#include "Exception.h"
+#include "XsensDataReader.h"
+
+namespace OpenSim {
+
+const std::string XsensDataReader::Orientations{ "orientations" };
+const std::string XsensDataReader::LinearAccelerations{ "linear_accelerations" };
+const std::string XsensDataReader::MagneticHeading{ "magnetic_heading" };
+const std::string XsensDataReader::AngularVelocity{ "angular_velocity" };
+
+XsensDataReader::XsensDataReader() {
+    constructProperties();
+}
+
+XsensDataReader::XsensDataReader(const std::string& aXmlFile) : Object (aXmlFile) {
+    constructProperties();
+    updateFromXMLDocument();
+}
+
+void
+XsensDataReader::constructProperties() {
+    constructProperty_data_folder("");
+    constructProperty_trial_prefix("");
+    constructProperty_ExperimentalSensors();
+}
+
+DataAdapter::OutputTables XsensDataReader::readTrial() {
+
+    std::vector<std::ifstream*> imuStreams;
+    std::vector<std::string> labels;
+    // files specified by prefix + file name exist
+    double dataRate = SimTK::NaN;
+    int packetCounterIndex = -1;
+    int accIndex = -1;
+    int gyroIndex = -1;
+    int magIndex = -1;
+    int rotationsIndex = -1;
+
+    int n_imus = getProperty_ExperimentalSensors().size();
+    int last_size = 1024;
+    // Will read data into pre-allocated Matrices in-memory rather than appendRow
+    // on the fly to avoid the overhead of 
+    SimTK::Matrix_<SimTK::Quaternion> rotationsData{ last_size, n_imus };
+    SimTK::Matrix_<SimTK::Vec3> linearAccelerationData{ last_size, n_imus };
+    SimTK::Matrix_<SimTK::Vec3> magneticHeadingData{ last_size, n_imus };
+    SimTK::Matrix_<SimTK::Vec3> angularVelocityData{ last_size, n_imus };
+    std::vector<double> times;
+    times.resize(last_size);
+    
+    std::string folderName = get_data_folder();
+    std::string prefix = get_trial_prefix();
+    for (int index = 0; index < n_imus; ++index) {
+        const ExperimentalSensor& nextItem = get_ExperimentalSensors(index);
+        auto fileName = folderName + prefix + nextItem.getName() +".txt";
+        auto* nextStream = new std::ifstream{ fileName };
+        OPENSIM_THROW_IF(!nextStream->good(),
+            FileDoesNotExist,
+            fileName);
+        // Add imu name to labels
+        labels.push_back(nextItem.get_name_in_model());
+        // Add corresponding stream to imuStreams
+        imuStreams.push_back(nextStream);
+
+        // Skip lines to get to data
```

Comment:
This seems unreliable. What if skipped rows is not 6 in other Xsens files and just 5? Can this be made user configurable. A default of 6 seems reasonable.

Reply:
@aseth1 Thanks for the feedback,  I refactored the code so that the number appears in exactly one place. I don't think this should be user configurable. If the format changes, likely other places will change as well (e.g. labels etc.).